odoo/tests/test-uninstalls.py

```diff
@@ -0,0 +1,135 @@
+            env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+            module = env['ir.module.module'].browse(module_id)
+            module.button_immediate_uninstall()
+    _logger.info('%s uninstalled', module_name)
+
+
+def cycle(db_name, module_id, module_name):
+    install(db_name, module_id, module_name)
+    uninstall(db_name, module_id, module_name)
+    install(db_name, module_id, module_name)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Script for testing the install / uninstall / reinstall cycle of Odoo modules")
+    parser.add_argument("--database", "-d", type=str, required=True,
+        help="The database to test (note: must have only 'base' installed)")
+    parser.add_argument("--skip", "-s", type=str, nargs=1,
+        help="Comma-separated list of modules to skip (they will only be installed)")
+    parser.add_argument("--resume", "-r", type=str, nargs=1,
+        help="Skip the cycle on all modules up to the specified one")
+    parser.add_argument("--addons-path", "-p", type=str, nargs=1,
+        help="Comma-separated list of paths to directories containing extra Odoo modules")
+    parser.add_argument("--uninstall", "-U", type=str,
+        help="Comma-separated list of modules to uninstall/reinstall")
+    return parser.parse_args()
+
+
+def test_full(args):
+    """ Test full install/uninstall/reinstall cycle for all modules """
+    with odoo.api.Environment.manage():
+        with odoo.registry(args.database).cursor() as cr:
+            env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+
+            def valid(module):
+                return not (
+                    module.name in BLACKLIST
+                    or module.name.startswith(IGNORE)
+                    or module.state in ('installed', 'uninstallable')
+                )
+
+            modules = env['ir.module.module'].search([]).filtered(valid)
+
+            # order modules in topological order
+            modules = modules.browse(topological_sort({
+                module.id: module.dependencies_id.depend_id.ids
+                for module in modules
+            }))
+            module_names = {module.id: module.name for module in modules}
+
+    resuming = args.resume
+    for module_id, module_name in module_names.items():
+        skip = False
+
+        # handle resume option
+        if resuming:
+            if module_names[module_id] == resuming[0]:
+                resuming = None
+            else:
+                skip = True
+
+        # handle skip option
+        if args.skip and module_name in args.skip[0].split(','):
+            skip = True
+
+        if skip:
+            install(args.database, module_id, module_name)
+        else:
+            cycle(args.database, module_id, module_name)
+
+
+def test_uninstall(args):
+    """ Tries to uninstall/reinstall one ore more modules"""
+    domain = [('name', 'in', args.uninstall.split(',')), ('state', '=', 'installed')]
```

Comment:
should we silently ignore modules that aren't installed? depends on the use case for this mode...

Reply:
The use case is for the runbot, it creates a database with everything installed and then create subbuilds that restaore this full database en tries to uninstall each module in a separate build.
So, yes, we silently ignore module that aren't installed.